### PR TITLE
Learning application catalogue

### PR DESCRIPTION
**Pull request by cliftonc**
Initial version of the API provided by the learning application for the catalogue.

@mcginkel have left a place holder for the sub-structure of the product we discussed yesterday (streams, assignments etc).

### CODE REVIEW

**Comment by cliftonc** [comment]
Preview here: https://stichtingsem.stoplight.io/docs/sem-technology-prototype/branches/learning-application-catalogue/reference/catalogue.v1.yaml

File: reference/catalogue.v1.yaml [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+  - url: 'https://example.stichtingsem.org/api'
+    description: catalogue
+paths:
+  /products:
+    get:
+      summary: Get Products from Catalogue
+      tags: []
+      parameters:
+        - in: query
+          name: subjectCode
```
Comment (mcginkel) [inline reference/catalogue.v1.yaml:30]: subjectcode is different for for the levels. The code for "English" is different for vmbo-gt-1 and havo-1. some times they share a code (havo-1, vwo-1) but i would place the subject under the level in the hierarchy

File: reference/catalogue.v1.yaml [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+              summary: The default value if none is provided
+            max:
+              value: 100
+              summary: The largest recommended page size
+        - schema:
+            type: string
+          in: query
+          name: status
+          description: |
+            Filter by status
```
Comment (mcginkel) [inline reference/catalogue.v1.yaml:72]: do we have a fixed set if statusses: add them in the doc

File: reference/catalogue.v1.yaml [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+      description: Get products from the catalogue - either all or those matching specific criteria.
+  '/product/{id}':
+    parameters:
+      - schema:
+          type: string
+        name: id
+        in: path
+        required: true
+    get:
+      summary: Get specific Product by ID
```
Comment (mcginkel) [inline reference/catalogue.v1.yaml:92]: why ID and not EAN?

File: reference/catalogue.v1.yaml [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+          description: |
+            An array of related product ID's (for display alongside a product in a marketplace).
+          items:
+            type: string
+        bundledProducts:
+          type: array
+          description: A link to any other products that are included in this one if it is a package / bundle.
+          items:
+            type: string
+      required:
```
Comment (mcginkel) [inline reference/catalogue.v1.yaml:197]: the property I miss is the list of streamcode that this product contains. If you "just" have the EAN the product is defined, but you have no clue what is in there.  It is the most important link to the content for reporting progress and meta data.

Often a EAN contains several streams ( example you can access material for next and previous year). The stream (or method) could also be part of another product, that is created to sell in a different way. 

There is an n-to-n relation between a product and a method.

Can we add: array of string: [ called streams or methods or streamcode]

File: reference/entitlement.v1.yaml [hunk 5]
```diff
@@ -285,44 +299,51 @@ components:
...
-        standard_subject_code:
+        subjectCode:
           type: string
-          description: Government standard subject code.
-        standard_level:
+          description: 'Officiële elementcode, bron: Ministerie van OCW'
+        level:
           type: string
-          description: Government standard level.
+          description: 'havo, vwo, etc.'
```
Comment (mcginkel) [inline reference/entitlement.v1.yaml:314]: should include level/year like havo-1, vwo-2, vmbo-gt-3

File: reference/catalogue.v1.yaml [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+          type: string
+          description: |
+            Status of product
+          enum:
+            - active
+            - not-for-sale
+        name:
+          type: string
+          description: A short name
+        subjectCodes:
```
[the remark this answers is not in the document]
Comment (cliftonc) [reply reference/catalogue.v1.yaml:134]: @niesink this should be aligned between MP / LMS / SIS and LMS - are you able to comment?   I think we really need to get the data structure right here, and make sure it works for all the different levels and combinations.  I'll also dig deeper on my side and see how we store this in our back end.

File: reference/catalogue.v1.yaml (hunk 3 above)
Comment (cliftonc) [reply reference/catalogue.v1.yaml:92]: Good spot - an oversight.

File: reference/catalogue.v1.yaml (hunk 4 above)
Comment (cliftonc) [reply reference/catalogue.v1.yaml:197]: I think we may be using different words for the same thing (I acknowledge that a parent `method` or equivalent is missing).  Can you provide an example - perhaps one for Noordhoff, Thieme and Malmberg so we all have the same understanding?

File: reference/entitlement.v1.yaml (hunk 5 above)
Comment (cliftonc) [reply reference/entitlement.v1.yaml:314]: Thanks

Comment (cliftonc) [reply reference/entitlement.v1.yaml:314]: Actually - just realised this was a rogue addition to my commit - its from entitlements, so will fix it there.

File: reference/catalogue.v1.yaml (hunk 1 above)
Comment (cliftonc) [reply reference/catalogue.v1.yaml:30]: In filtering I assume you may search for just one or the other?  

`?level=havo-1`
`?subjectCode=math-subject-code&level=havo-1`

I think the question is if you create a tuple with { level, subjectCode } - do you always have to filter by both?

File: reference/catalogue.v1.yaml (hunk 4 above)
Comment (cliftonc) [reply reference/catalogue.v1.yaml:199]: I've added the object as discussed, but am need to check internally if it makes sense from our perspective - I don't feel we model it this way, but can easily be wrong :D

File: reference/catalogue.v1.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (cliftonc) [reply reference/catalogue.v1.yaml]: I've created an issue to discuss: https://github.com/stichtingsem/technology-prototype/issues/39

File: reference/catalogue.v1.yaml (hunk 1 above)
Comment (cliftonc) [reply reference/catalogue.v1.yaml:30]: I've added an issue to discuss: https://github.com/stichtingsem/technology-prototype/issues/39

**Comment by cliftonc** [comment]
@mcginkel @dpronk are you able to look at this today?  I'm keen to get a baseline merged (there will be further iterations and additions as more people look at it I'm sure).

**Comment by dpronk** [comment]
For sure 

Verstuurd vanaf mijn iPhone

**Comment by cliftonc** [comment]
I'm keen to merge this as is - as it allows others to view it at the meeting tomorrow, @mcginkel is that OK with you?

**Comment by cliftonc** [comment]
Bump ... I'm keen to merge this as is - as it allows others to view it at the up-coming meetings, @mcginkel is that OK with you?

**Comment by cliftonc** [comment]
Merging - just to get this initial version in for future work.

**Pull request merged**